### PR TITLE
auto disable shadow when blur is enabled and make query window width dynamic

### LINKED ISSUE #478: query window when shadow and blur effects enabled, extends outside of the window
currently in dev branch as at commit [2e412a00232998546f987c034adc753c59c36b06](https://github.com/Flow-Launcher/Flow.Launcher/tree/2e412a00232998546f987c034adc753c59c36b06), query window when blur enabled via the theme, extends outside of the window

repro:

1. enable shadow
2. switch to BlurBlack Darker theme
3. restart flow
4. the edges where there should be shadow becomes just blurred:
image

### PR DESCRIPTION
**Pull request by jjw24**
- shadow does not work when blur is enabled via the theme file. This automatically disables shadow when blur is enabled
close #478 

- make query window width dynamic

### CODE REVIEW

**Comment by taooceros** [comment]
Why not enable Blur when blur theme is selected instead of prompting to restart?

**Comment by jjw24** [comment]
Can't enable it with the current implementation of blur, you can call the set blur function but it comes out with wider blur width, hence the original comment about it being in mainwindow's on_loaded function call

**Comment by taooceros** [comment]
> Can't enable it with the current implementation of blur, you can call the set blur function but it comes out with wider blur width, hence the original comment about it being in mainwindow's on_loaded function call

I know, but the extra boarder will only be enabled when shadow effect is on. I mean if that's not on, we can enable the blur effect so that user won't see a blur theme without blur effect.

**Comment by jjw24** [comment]
cant, blur can only be enabled by mainwindow's onloaded, if you force it else where you get this

image

**Comment by taooceros** [comment]
I take a debug check. It is behaving weird if we set blur after onload because the width of the main window is changed to 1440 instead of expected 750 with unknown reason. If we change it back to 750 (or the dynamic width setting in #374), it seems work correctly.

**Comment by jjw24** [comment]
oh right, good pick up thank you.

**Comment by taooceros** [comment]
```c#
private void SetWindowAccent(Window w, AccentState state)
        {
            var windowHelper = new WindowInteropHelper(w);
            w.Width = 750;
            windowHelper.EnsureHandle();
            var accent = new AccentPolicy { AccentState = state };
            var accentStructSize = Marshal.SizeOf(accent);

            var accentPtr = Marshal.AllocHGlobal(accentStructSize);
            Marshal.StructureToPtr(accent, accentPtr, false);

            var data = new WindowCompositionAttributeData
            {
                Attribute = WindowCompositionAttribute.WCA_ACCENT_POLICY,
                SizeOfData = accentStructSize,
                Data = accentPtr
            };

            SetWindowCompositionAttribute(windowHelper.Handle, ref data);

            Marshal.FreeHGlobal(accentPtr);
        }
```

This will work, because the setting window will hide the main window, so we need to ensure the mainwindow is rendered successfully before assigning it with a shadow effect.

**Comment by jjw24** [comment]
Let's make that width dynamically passed in from theme xmal, should work well since we are making width dynamic in the other pr as well

**Comment by jjw24** [comment]
all done

File: Flow.Launcher.Core/Resource/Theme.cs [hunk 1]
```diff
@@ -320,35 +343,39 @@ private enum WindowCompositionAttribute
...
         public void SetBlurForWindow()
         {
+            if (BlurEnabled)
+            {
+                SetWindowAccent(Application.Current.MainWindow, AccentState.ACCENT_ENABLE_BLURBEHIND);
+            }
+            else
+            {
+                SetWindowAccent(Application.Current.MainWindow, AccentState.ACCENT_DISABLED);
+            }
```
Comment (taooceros) [inline Flow.Launcher.Core/Resource/Theme.cs:353]: Maybe only change this if blur state is changed?

Comment (jjw24) [reply Flow.Launcher.Core/Resource/Theme.cs:353]: in the original code this state is always set though via on_loaded everytime when flow is started, not sure if we really need to manage this

Comment (jjw24) [reply Flow.Launcher.Core/Resource/Theme.cs:353]: @taooceros did you want to show me what you mean?

Comment (taooceros) [reply Flow.Launcher.Core/Resource/Theme.cs:353]: Oh sorry I forget to reply. I mean only call this method when we switch from Blur to non-blur or non-blur to Blur. Also maybe remove the call of this method in MainWindow.OnLoaded and move that to ChangeTheme

Comment (jjw24) [reply Flow.Launcher.Core/Resource/Theme.cs:353]: > call this method when we switch from Blur to non-blur or non-blur to Blur

there's not really a good way to tell what state blur was previously. is this just so the call is not done multiple times when user cycles through themes? i tested current way there hasnt been any issues.

>  remove the call of this method in MainWindow.OnLoaded and move that to ChangeTheme

this is already removed.

**Review by taooceros — APPROVED** [review]

**Pull request merged**